### DISCUSSION TITLE
Unpack library

### DISCUSSION THREAD

**Discussion by codito** [Ephemerality/xray-builder.gui#43] [Ideas]
Thank you for the amazing work in this code base and making it open/public. I'm working on a opensource [tool][] to extract kindle highlights along with context from the ebook/document, and was able to successfully use the [unpack][] portion of this code base. There is a lack of good `mobi` or `azw3` parsers for .net, above code can definitely fill the gap.

Is there a desire/possibility to release the [unpack][] portion as a separate library?

[tool]: https://github.com/codito/noted/
[unpack]: https://github.com/Ephemerality/xray-builder.gui/tree/master/XRayBuilder.Core/src/Unpack

**Comment by Ephemerality** [comment]
I could probably do that, yeah. It could probably use some cleaning up first though, it was written with no real design in mind and mostly just suits my needs. It also handles KFX files if you only care about the text content and not the full markup.

**Reply by codito** [reply]
Awesome! Looking forward to test it out.

**Reply by Ephemerality** [reply]
I started trying to get it set up to put on NuGet but I forgot that my KFX support requires a modified version of Amazon's IonDotNet library (forked here: https://github.com/Ephemerality/ion-dotnet) which I probably can't get away with putting on NuGet as well.
Would it work if I made the unpack portion into a separate github repo that you could use via `git submodule`?

**Reply by codito** [reply]
`git submodule` will add a coupling with source code, which may block refactoring. May I suggest one more approach, we can evaluate and decide.

Would it work if we publish the modified ion-dotnet to a private feed e.g. github packages or myget, and package the dll as a private resource in unpack's nuget package? It may help us keep the dependencies to binary level and not depend on source. Consumers will not be aware of this dependency unless they also try using the public ion-dotnet package, but that may be acceptable given the dependency on private fork. Someday, when the public ion-dotnet has the necessary capability, we can transparently drop the private packaging and simply add a Package Dependency in unpack's nuspec.

We've tried a similar approach at https://github.com/spekt/. The `testlogger` library is a private library that gets packaged in `nunitxml.testlogger` and other public packages. Few references: https://github.com/spekt/nunit.testlogger/blob/master/NuGet.config#L5, https://github.com/spekt/nunit.testlogger/blob/master/src/package/NUnitXml.TestLogger.nuspec#L17 

I can try to set it up for unpack or create a PR if this sounds interesting.

**Reply by Ephemerality** [reply]
Ah, that seems like a good option. I'll give it a go and let you know if I need help with it. Thanks!

**Reply by Ephemerality** [reply]
I tried popping it into GitHub Packages, referencing that in the Unpack project (using the nuget.config reference to github package) and then sending that off to NuGet, but NuGet thought that I was referring to an `Amazon.IonDotNet.Ephemerality` package on NuGet instead of on GitHub.  
Might've missed a step, not sure. However I did just pop the .dll into the Unpack project instead and sent that to nuget so in theory you should be able to add `<PackageReference Include="Ephemerality.Unpack" Version="0.0.2" />` and have it work, that's a bit icky possibly...  

I haven't updated the readme yet, but I did some renaming so if you want to open a .mobi or .azw3 it's just `new MobiMetadata(stream)`.
For KFX you can do `new KfxContainer(stream)`. The KFX implementation doesn't support `GetRawmlStream` though, to get the book's content (text only, no markup), I currently do this:  
```
kfx.GetContentChunks()
    .Where(contentChunk => contentChunk.ContentText != null)
    .Select(contentChunk => new Paragraph
    {
        ContentText = contentChunk.ContentText,
        Location = contentChunk.Pid
    });
```  
The KFX support is also not 100% implemented either, I didn't have enough example books to test.  
  
Let me know if that works for you or if I need to do anything else

Edit: I guess the step I missed was the whole .nuspec file referencing the DLL from somewhere... not sure how to set that up. The Ion package is on GitHub Packages @ `https://nuget.pkg.github.com/Ephemerality/index.json`.

**Reply by codito** [reply]
Thank you, this is great news! Mobi functionality works flawlessly. Will try out kfx in the weekend and report back.

The nuget package install confusion may be because Github packages do not support anonymous access: https://docs.github.com/en/packages/guides/configuring-dotnet-cli-for-use-with-github-packages#authenticating-to-github-packages or the `nuget.config` file is not pointing to the private feed. For auth problem, myget might be a better option. To let nuget search for the custom ion package in the private feed, a nuget.config needs to be added at project root similar to https://github.com/spekt/nunit.testlogger/blob/master/NuGet.config

**Reply by codito** [reply]
@Ephemerality I finally got a chance to try out Kfx (sorry for the long delay). The IonDotnet assembly was not getting packaged in nuget package due to nuget bug, added an workaround and created https://github.com/Ephemerality/xray-builder.gui/pull/44. We need to do a version bump to `0.0.3`. I can add it to the PR if it looks okay.

On local validation, I am able to instantiate Kfx reader, but hitting DRM blocker exception. It appears to be by design. Please suggest if there's a way to get a DRM free kfx to test out.

**Reply by Ephemerality** [reply]
Ah that's interesting. Yeah you can add the 0.0.3 bump to the PR, then I'll merge it & get it up on NuGet.  
  
You're right, only non-DRM'd books are supported (breaking DRM is out of scope of the project but doable with another tool). 
 Here's a public domain KFX to try (converted with JHowell's KFX output plugin):
[Frankenstein - Mary W. Shelley.kfx.zip](https://github.com/Ephemerality/xray-builder.gui/files/6299017/Frankenstein.-.Mary.W.Shelley.kfx.zip)

**Reply by codito** [reply]
Perfect, KFX is working now, thank you. Updated #44 with a version bump.

**Reply by Ephemerality** [reply]
Sorry for the delay, 0.0.3 should be available on NuGet now.
